### ISSUE TITLE
Accessing global variables inside a macro

### ISSUE THREAD

**Issue by electrotype** [PebbleTemplates/pebble#231]
I [see ](http://www.mitchellbosecke.com/pebble/documentation/tag/macro):

> A macro does not have access to the same variables that the rest of the template has access to. A macro can only work with the variables provided as arguments.

But is there a way to _pass_ the global scope to a macro?

In one macro, I call a function which needs to access some variables which are on the global context, but I don't know the keys at the time I call the macro. 

For example : 

```
{% macro myMacro(someList %}
    {% for element in someList%}
        {{displayFromKey(element.key)}}  
    {% endfor %}
{% endmacro %}
```

The `displayFromKey()` function needs to use the key of an element to get the associated variable from the global context. Something like :

```
// inside displayFromKey() :
String someData = ((EvaluationContext)args.get("_context")).getScopeChain().get(args.get("elementKey"));
```

But it seems the global context is not available inside `displayFromKey()` because it is called from a macro. Is there a workaround that could help me?

**Comment by electrotype** [comment]
I guess the solution is to `include` another template instead of using a macro. This works because the included template has access to the same variables than the parent one.

But I wish I could define a template fragment in the same page as the including template, as you can do with a macro.

**Comment by mbosecke** [comment]
I have completely removed the restriction on macros that was preventing them from access all global variables; implemented on the 3.0 branch. 

Honestly I don't really know why i added that restriction in the first place. I'm hoping there aren't any consequences to this change that I just haven't thought of yet.

**Comment by setevoy** [comment]
Seems like related issue.
I use template block "info" to store some metadata (yaml) inside template. I evaluate it during site load, outside of any HTTP request. 
For some reason when I'm evaluating info block, rest of template is evaluated also (BTW why?). So I setting "loading" flag to indicate, it's not regular request. I wrote extension functions and they use "loading" flag to return reasonable defaults instead of trying to access HTTP request.
It worked until I called one of such functions from macro. Now function doesn't see my flag. Here is test case:

```Java
    @Test
    public void noAccessToContextFromMacro() throws PebbleException, IOException {
        PebbleEngine pebble = new PebbleEngine.Builder().loader(new StringLoader())
                .extension(new AbstractExtension() {
                    @Override
                    public Map<String, Function> getFunctions() {
                        return Collections.singletonMap("func", new Function() {
                            @Override
                            public Object execute(Map<String, Object> args, PebbleTemplate self, EvaluationContext context, int lineNumber) {
                                if (context.getVariable("loading") != null) return null;
                                throw new IllegalStateException("NO HTTP REQUEST!");
                            }

                            @Override
                            public List<String> getArgumentNames() {
                                return Collections.EMPTY_LIST;
                            }
                        });
                    }
                })
                .strictVariables(false).build();
        String templateContent = "{% block info %}someinfo{% endblock %}{{ test(_context) }}{% macro test(_context) %}{{ func() }}{% endmacro %}";
        PebbleTemplate template = pebble.getTemplate(templateContent);

        Writer writer = new StringWriter();
        template.evaluateBlock("info", writer, Collections.singletonMap("loading", true));

        assertEquals("someinfo", writer.toString());
    }

```

I'm not sure, but seems like && should be instead of || here: https://github.com/PebbleTemplates/pebble/blob/5f8d9ccc3639f7872c8c0df7505491e1aa5f5575/pebble/src/main/java/com/mitchellbosecke/pebble/template/ScopeChain.java#L133
